### PR TITLE
Sync engagement database -> Rapid Pro

### PR DESCRIPTION
**Pull request by as2388**
Prototype implementation of sync from an engagement database to a Rapid Pro workspace. This syncs normal fields only, I'll add support for setting a consent_withdrawn field in another PR.

Note that this is designing for two modes of operation - a realtime sync, probably for consent only, and as one-off manual tool runs, e.g. for syncing the database to a new workspace or for when we've marked data as stale every so often.

### CODE REVIEW

File: src/engagement_db_to_rapid_pro/engagement_db_to_rapid_pro.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+            messages = []
+            for dataset in dataset_config.engagement_db_datasets:
+                messages.extend(datasets.get(dataset, []))
+
+            # Only overwrite this contact field if there is data to write or it's ok to clear a field.
+            if len(messages) > 0:
+                if sync_config.write_mode == WriteModes.SHOW_PRESENCE:
+                    contact_fields[dataset_config.rapid_pro_contact_field] = _PRESENCE_VALUE
+                else:
+                    assert sync_config.write_mode == WriteModes.CONCATENATE_TEXTS
```
Comment (IsaackMwenda) [inline src/engagement_db_to_rapid_pro/engagement_db_to_rapid_pro.py:71]: Which instance would we want to concatenate texts in rapidpro.

File: src/engagement_db_to_rapid_pro/engagement_db_to_rapid_pro.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+        for dataset_config in sync_config.normal_datasets:
+            # Find all the messages from this participant that are relevant to this dataset.
+            messages = []
+            for dataset in dataset_config.engagement_db_datasets:
+                messages.extend(datasets.get(dataset, []))
+
+            # Only overwrite this contact field if there is data to write or it's ok to clear a field.
+            if len(messages) > 0:
+                if sync_config.write_mode == WriteModes.SHOW_PRESENCE:
+                    contact_fields[dataset_config.rapid_pro_contact_field] = _PRESENCE_VALUE
```
Comment (IsaackMwenda) [inline src/engagement_db_to_rapid_pro/engagement_db_to_rapid_pro.py:69]: We might need to update rapid_pro -> db sync not to add _PRESENCE_VALUE to DB. The alternative is to update the field used to check whether the contact has sent a response before e.g `age_complete`

File: src/engagement_db_to_rapid_pro/engagement_db_to_rapid_pro.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+                else:
+                    assert sync_config.write_mode == WriteModes.CONCATENATE_TEXTS
+                    contact_fields[dataset_config.rapid_pro_contact_field] = ";".join([m.text for m in messages])
+            elif sync_config.allow_clearing_fields:
+                contact_fields[dataset_config.rapid_pro_contact_field] = ""
+
+        # TODO: Detect and update consent status
+
```
Comment (IsaackMwenda) [inline src/engagement_db_to_rapid_pro/engagement_db_to_rapid_pro.py:77]:
```suggestion
# TODO: Update special group membership status e.g listening groups
```

File: src/engagement_db_to_rapid_pro/engagement_db_to_rapid_pro.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+        if msg.dataset not in participant:
+            participant[msg.dataset] = []
+        dataset = participant[msg.dataset]
+
+        dataset.append(msg)
+
+    # Make sure all the contact fields exist in the Rapid Pro workspace.
+    existing_contact_fields = [f.key for f in rapid_pro.get_fields()]
+    contact_fields_to_sync = [dataset_config.rapid_pro_contact_field for dataset_config in sync_config.normal_datasets]
+    for contact_field in contact_fields_to_sync:
```
Comment (IsaackMwenda) [inline src/engagement_db_to_rapid_pro/engagement_db_to_rapid_pro.py:48]: The ability to switch between live and dry runs would be great for this.

**Review by IsaackMwenda — APPROVED** [review]

File: src/engagement_db_to_rapid_pro/engagement_db_to_rapid_pro.py (hunk 1 above)
Comment (as2388) [reply src/engagement_db_to_rapid_pro/engagement_db_to_rapid_pro.py:71]: I think it's a useful debug tool; it's much easier to understand/check what's going on in rapid pro if you can see what the messages were

File: src/engagement_db_to_rapid_pro/engagement_db_to_rapid_pro.py (hunk 2 above)
Comment (as2388) [reply src/engagement_db_to_rapid_pro/engagement_db_to_rapid_pro.py:69]: I think all we care about in Rapid Pro is presence (for Rapid Pro to decide if it needs to ask a particular question), we can treat everything in rapid pro as a "complete" field

File: src/engagement_db_to_rapid_pro/engagement_db_to_rapid_pro.py (hunk 4 above)
Comment (as2388) [reply src/engagement_db_to_rapid_pro/engagement_db_to_rapid_pro.py:48]: Agreed, I'll add dry runs to the task list to be completed later

File: src/engagement_db_to_rapid_pro/engagement_db_to_rapid_pro.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+                if sync_config.write_mode == WriteModes.SHOW_PRESENCE:
+                    contact_fields[dataset_config.rapid_pro_contact_field] = _PRESENCE_VALUE
+                else:
+                    assert sync_config.write_mode == WriteModes.CONCATENATE_TEXTS
+                    contact_fields[dataset_config.rapid_pro_contact_field] = ";".join([m.text for m in messages])
+            elif sync_config.allow_clearing_fields:
+                contact_fields[dataset_config.rapid_pro_contact_field] = ""
+
+        # TODO: Detect and update consent status
+# TODO: Update special group membership status e.g listening groups
```
Comment (DanielMwendwa) [inline src/engagement_db_to_rapid_pro/engagement_db_to_rapid_pro.py:77]: format

**Review by DanielMwendwa — APPROVED** [review]

File: src/engagement_db_to_rapid_pro/engagement_db_to_rapid_pro.py (hunk 1 above)
Comment (IsaackMwenda) [reply src/engagement_db_to_rapid_pro/engagement_db_to_rapid_pro.py:71]: Maybe am missing something here - but I think we should not be updating texts in fields that also store direct responses from users. e.g storing age response from Imaqal to a csap age responce field.

Comment (as2388) [reply src/engagement_db_to_rapid_pro/engagement_db_to_rapid_pro.py:71]: resolved offline to include origins in contact field write-backs in a future pr

**Pull request merged**